resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/submitsteps/hadoopsteps/HadoopKerberosKeytabResolverStep.scala

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.k8s.submit.submitsteps.hadoopsteps
+
+import java.io._
+import java.security.PrivilegedExceptionAction
+
+import scala.collection.JavaConverters._
+
+import io.fabric8.kubernetes.api.model.SecretBuilder
+import org.apache.commons.codec.binary.Base64
+import org.apache.hadoop.fs.FileSystem
+import org.apache.hadoop.security.Credentials
+import org.apache.hadoop.security.token.{Token, TokenIdentifier}
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.deploy.k8s.{HadoopUGIUtil, KerberosTokenConfBootstrapImpl, PodWithMainContainer}
+import org.apache.spark.deploy.k8s.constants._
+import org.apache.spark.internal.Logging
+
+ /**
+  * This step does all the heavy lifting for Delegation Token logic. This step
+  * assumes that the job user has either specified a principal and keytab or ran
+  * $kinit before running spark-submit. With a TGT stored locally, by running
+  * UGI.getCurrentUser you are able to obtain the current user, alternatively
+  * you can run UGI.loginUserFromKeytabAndReturnUGI and by running .doAs run
+  * as the logged into user instead of the current user. With the Job User principal
+  * you then retrieve the delegation token from the NameNode and store values in
+  * DelegationToken. Lastly, the class puts the data into a secret. All this is
+  * appended to the current HadoopSpec which in turn will append to the current
+  * DriverSpec.
+  */
+private[spark] class HadoopKerberosKeytabResolverStep(
+  submissionSparkConf: SparkConf,
+  maybePrincipal: Option[String],
+  maybeKeytab: Option[File],
+  maybeRenewerPrincipal: Option[String],
+  hadoopUGI: HadoopUGIUtil) extends HadoopConfigurationStep with Logging {
+    private var originalCredentials: Credentials = _
```

Comment:
Avoid using `var` in general

Reply:
These are Hadoop objects, in java, that are being modified, I believe that I need var.

Reply:
We only have one method in this class - can't all of the fields be defined as `vals` as they are being created?